File: gpt-3.5-ft-large/context/loop_construct_reduction_clause.c
#ifndef T1
//T1:reduction,loop,combined-constructs,V:2.0-2.7
int test1(){
    int err = 0;
    srand(SEED);
    real_t * a = (real_t *)malloc(10 * n * sizeof(real_t));
    real_t * b = (real_t *)malloc(10 * n * sizeof(real_t));
    real_t avg = 0.0;

    for (int x = 0; x < 10 * n; ++x){
        a[x] = rand() / (real_t)(RAND_MAX / 10);
        b[x] = rand() / (real_t)(RAND_MAX / 10);
    }

    #pragma acc data copyin(a[0:10*n], b[0:10*n]) copy(avg)
    {
        #pragma acc parallel loop reduction(+:avg)
        for (int x = 0; x < 10 * n; ++x){
            avg += a[x] * b[x];
        }
    }
    avg = avg / (n * 10);
    real_t sum = 0.0;

    for (int x = 0; x < 10 * n; ++x){
        sum += a[x] * b[x];
    }
    sum = sum / (n * 10);

    if (fabs(sum - avg) > (2.0 / n) && sum != avg){
        err = 1;
    }

    return err;
}
#endif

#ifndef T2
//T2:reduction,loop,combined-constructs,V:2.0-2.7
int test2(){
    int err = 0;
    srand(SEED);
    real_t * a = (real_t *)malloc(25 * n * sizeof(real_t));
    real_t * b = (real_t *)malloc(25 * n * sizeof(real_t));
    real_t avg[5];

    for (int x = 0; x < 25 * n; ++x){
        a[x] = 1;
        b[x] = 1;
    }

    for (int x = 0; x < 5; ++x){
        avg[x] = 0.0;
    }

    #pragma acc data copyin(a[0:25*n], b[0:25*n]) copy(avg[0:5])
    {
        #pragma acc parallel loop num_gangs(5) reduction(+:avg)
        for (int x = 0; x < 5 * n; ++x){
            avg[x%5] += a[x] * b[x];
        }
    }

    for (int x = 0; x < 25 * n; ++x){
        if (fabs(a[x]*b[x] - (1 + avg[x%5])) > PRECISION){
            err = 1;
        }
    }

    return err;
}
#endif

#ifndef T3
//T3:reduction,loop,combined-constructs,V:2.0-2.7
int test3(){
    int err = 0;
    srand(SEED);
    real_t * a = (real_t *)malloc(25 * n * sizeof(real_t));
    real_t * b = (real_t *)malloc(25 * n * sizeof(real_t));
    real_t avg[5];

    for (int x = 0; x < 25 * n; ++x){
        a[x] = 1;
        b[x] = 1;
    }

    for (int x = 0; x < 5; ++x){
        avg[x] = 0.0;
    }

    #pragma acc data copyin(a[0:25*n], b[0:25*n]) copy(avg[0:5])
    {
        #pragma acc parallel loop reduction(+:avg)
        for (int x = 0; x < 5 * n; ++x){
            avg[x%5] += a[x] * b[x];
        }
    }

    for (int x = 0; x < 5; ++x){
        avg[x] = avg[x] / 5 / n;
    }

    for (int x = 0; x < 5 * n; ++x){
        if (fabs(avg[x%5] - 1) > PRECISION){
            err = 1;
        }
    }

    return err;
}
#endif

int main(){
    int failcode = 0;
    int failed;
#ifndef T1
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test1();
    }
    if (failed != 0){
        failcode = failcode + (1 << 0);
    }
#endif
#ifndef T2
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test2();
    }
    if (failed != 0){
        failcode = failcode + (1 << 1);
    }
#endif
#ifndef T3
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test3();
    }
    if (failed != 0){
        failcode = failcode + (1 << 2);
    }
#endif
    return failcode;
}

File: gpt-3.5-finetuned/stage2/loop_construct_reduction_clause.f90
!T1:loop construct reduction clause,V:2.7-2.3
      LOGICAL FUNCTION test1()
        USE OPENACC
        IMPLICIT NONE
        INCLUDE "acc_testsuite.Fh"
        
        INTEGER :: errors = 0

        REAL(8),DIMENSION(LOOPCOUNT, 10):: a, b, extra
        REAL(8),DIMENSION(10):: total, totals
        INTEGER :: x, y, z, current_total

        SEEDDIM(1) = 1
        SEEDDIM(1) = SEED
        CALL RANDOM_SEED(PUT=SEEDDIM)

        CALL RANDOM_NUMBER(a)
        CALL RANDOM_NUMBER(b)
        CALL RANDOM_NUMBER(extra)

        total = 0
        totals = 0

        !$acc data copyin(a(1:LOOPCOUNT, 1:10), b(1:LOOPCOUNT, 1:10), extra(1:LOOPCOUNT, 1:10)) copy(total(1:10)) copyout(totals(1:LOOPCOUNT))
          !$acc parallel
            !$acc loop reduction(+:total(1:10))
            DO x = 1, LOOPCOUNT
              DO y = 1, 10
                total(y) = total(y) + a(x, y) * b(x, y)
              END DO
            END DO
            !$acc end loop
          !$acc end parallel
        !$acc end data

        DO x = 1, LOOPCOUNT
          DO y = 1, 10
            totals(x) = totals(x) + a(x, y) * b(x, y)
          END DO
        END DO

        DO x = 1, 10
          IF (totals(x) - total(x) .gt. PRECISION) THEN
            errors = errors + 1
          END IF
        END DO

        !$acc data copyin(a(1:LOOPCOUNT, 1:10), b(1:LOOPCOUNT, 1:10)) copy(total(1:10))
          !$acc parallel
            !$acc loop reduction(+:totals(1:LOOPCOUNT))
            DO x = 1, 10
              DO y = 1, LOOPCOUNT
                totals(y) = totals(y) + a(y, x) * b(y, x)
              END DO
            END DO
            !$acc end loop
          !$acc end parallel
        !$acc end data

        DO x = 1, 10
          current_total = 0
          DO y = 1, LOOPCOUNT
            current_total = current_total + a(y, x) * b(y, x)
          END DO
          IF (current_total - totals(x) .gt. PRECISION) THEN
            errors = errors + 1
          END IF
        END DO

        !$acc data copyin(a(1:LOOPCOUNT, 1:10), b(1:LOOPCOUNT, 1:10), extra(1:LOOPCOUNT, 1:10)) copy(total(1:10))
          !$acc parallel
            !$acc loop reduction(*:totals(1:LOOPCOUNT))
            DO x = 1, 10
              DO y = 1, LOOPCOUNT
                totals(y) = totals(y) * (a(y, x) + b(y, x) + extra(y, x))
              END DO
            END DO
            !$acc end loop
          !$acc end parallel
        !$acc end data

        DO x = 1, 10
          current_total = 1
          DO y = 1, LOOPCOUNT
            current_total = current_total * (a(y, x) + b(y, x) + extra(y, x))
          END DO
          IF (current_total - totals(x) .gt. PRECISION) THEN
            errors = errors + 1
          END IF
        END DO

        !$acc data copyin(a(1:LOOPCOUNT, 1:10), b(1:LOOPCOUNT, 1:10)) copy(total(1:10))
          !$acc parallel
            !$acc loop reduction(-:totals(1:LOOPCOUNT))
            DO x = 1, 10
              DO y = 1, LOOPCOUNT
                totals(y) = totals(y) - (a(y, x) + b(y, x))
              END DO
            END DO
            !$acc end loop
          !$acc end parallel
        !$acc end data

        DO x = 1, 10
          current_total = 0
          DO y = 1, LOOPCOUNT
            current_total = current_total - (a(y, x) + b(y, x))
          END DO
          IF (current_total - totals(x) .gt. PRECISION) THEN
            errors = errors + 1
          END IF
        END DO

        IF (errors .eq. 0) THEN
          test1 = .FALSE.
        ELSE
          test1 = .TRUE.
        END IF
      END

      PROGRAM main
        IMPLICIT NONE
        INTEGER :: failcode, testrun
        LOGICAL :: failed
        INCLUDE "acc_testsuite.Fh"
        LOGICAL :: test1
        failed = .FALSE.
        failcode = 0
        DO testrun = 1, NUM_TEST_CALLS
          failed = failed .or. test1()
        END DO
        IF (failed) THEN
          failcode = failcode + 2 ** 0
          failed = .FALSE.
        END IF
        CALL EXIT (failcode)
      END PROGRAM